language/move-prover/bytecode/src/escape_analysis.rs

```diff
@@ -0,0 +1,419 @@
+    cmp::Ordering,
+    collections::{BTreeMap, BTreeSet},
+};
+
+// =================================================================================================
+// Data Model
+
+#[derive(Debug, Clone, Copy, Eq, PartialEq)]
+pub enum AbsValue {
+    NonRef,
+    OkRef,
+    InternalRef,
+}
+
+impl AbsValue {
+    pub fn is_internal_ref(&self) -> bool {
+        matches!(self, Self::InternalRef)
+    }
+}
+
+type EscapeAnalysisState = MapDomain<TempIndex, AbsValue>;
+
+impl EscapeAnalysisState {
+    fn get_local_index(&self, i: &TempIndex) -> &AbsValue {
+        self.get(i)
+            .unwrap_or_else(|| panic!("Unbound local index {} in state {:?}", i, self))
+    }
+
+    fn assign(&mut self, lhs: TempIndex, rhs: &TempIndex) {
+        let rhs_value = *self.get_local_index(rhs);
+        self.insert(lhs, rhs_value);
+    }
+
+    pub fn call(&mut self, rets: &Vec<TempIndex>, args: &Vec<TempIndex>, call_env: &FunctionEnv) {
+        let has_internal_ref_input = args
+            .iter()
+            .any(|arg_index| self.get(&arg_index).unwrap().is_internal_ref());
+        for (ret_index, ret_type) in call_env.get_return_types().iter().enumerate() {
+            let ret_value = if ret_type.is_reference() {
+                if has_internal_ref_input {
+                    AbsValue::InternalRef
+                } else {
+                    AbsValue::OkRef
+                }
+            } else {
+                AbsValue::NonRef
+            };
+            self.insert(rets[ret_index], ret_value);
+        }
+    }
+}
+
+// =================================================================================================
+// Joins
+
+impl PartialOrd for AbsValue {
+    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
+        if self == other {
+            return Some(Ordering::Equal);
+        }
+        match (self, other) {
+            (_, AbsValue::InternalRef) => Some(Ordering::Less),
+            _ => None,
+        }
+    }
+}
+
+impl AbstractDomain for AbsValue {
+    fn join(&mut self, other: &Self) -> JoinResult {
+        if self == other {
+            return JoinResult::Unchanged;
+        }
+        // unequal; use top value
+        *self = AbsValue::InternalRef;
+        JoinResult::Changed
+    }
+}
+
+// =================================================================================================
+// Transfer functions
+
+#[derive(PartialOrd, PartialEq, Eq, Ord)]
+struct WarningId {
+    ret_index: usize,
+    offset: CodeOffset,
+}
+
+struct SpecMemoryInfo {
+    /// Fields that occur in struct, module, or global specs. Leaked references to fields inside
+    /// this set will be flagged, leaked references to other fields will be allowed.
+    relevant_fields: BTreeSet<(QualifiedId<StructId>, FieldId)>,
+    /// Structs that occur in struct, module, or global specs. Leaked references to fields inside
+    /// these structs may cause a spec like `invariant forall s: S: s == S { f: 10 }` to be false
```

Comment:
Do you mean a data (struct) invariant here instead of a global invariant? 

I interpret this PR as follows. We want to flag in which cases we return mut refs to callers which we consider untrusted, and which would need invariants to be asserted when mutation ends.

If that is the case, then only data invariants are relevant. Other specs cannot be deferred to the caller (for global invariants, we flag already an error if they are attempting to suspend invariants).

If I do not misunderstand the goal here, then the implementation could be much simpler. Go over all functions in the global env such that `fun.has_unknown_callers()` and check whether they take or return a `&mut T`. Then check whether there is any `T' in T` s.t. `T'` is in this module and `struct_env[T'].has_conditions()`. This would be based on the following observation: whenever there is a return `r: &mut R` then there is also an input parameter `x: &mut X` such that there is a selection path `s --> r`.

Reply:
> If I do not misunderstand the goal here, then the implementation could be much simpler. Go over all functions in the global env such that fun.has_unknown_callers() and check whether they take or return a &mut T. Then check whether there is any T' in T s.t. T' is in this module and struct_env[T'].has_conditions(). This would be based on the following observation: whenever there is a return r: &mut R then there is also an input parameter x: &mut X such that there is a selection path x --> r.

I think that this scheme is also sound, but is less precise. E.g., I think it would incorrectly flag the function `ok` in:

```
module M {
  struct S { f: u64 }
  spec S { invariant f >= 1 }

  public fun ok(x: &mut u64, s: &mut S): &mut u64 {
    s.f += 1;
    x
  }
```

I don't think this precision is likely to matter much in practice, but I'm somewhat attached to the current approach because we've already formalized it/implemented it/proven it correct + it's also pretty simple.

> If that is the case, then only data invariants are relevant. Other specs cannot be deferred to the caller (for global invariants, we flag already an error if they are attempting to suspend invariants).

Yep, I think that makes sense for the current vocabulary of global invariants. To make sure I understand: if we enrich the vocabulary to allow things like `invariant forall s: S: s == S { f: 10 }` (which is currently a dead invariant as we discussed in https://github.com/diem/diem/issues/9167, but you could imagine supporting it with the natural meaning), or other constructs that quantify over all fields of a given type (e.g., the classic "sum of all `value` fields of `Diem<T>` idea), we would then need to consider global invariants as relevant, right?

Reply:
> > If I do not misunderstand the goal here, then the implementation could be much simpler. Go over all functions in the global env such that fun.has_unknown_callers() and check whether they take or return a &mut T. Then check whether there is any T' in T s.t. T' is in this module and struct_env[T'].has_conditions(). This would be based on the following observation: whenever there is a return r: &mut R then there is also an input parameter x: &mut X such that there is a selection path x --> r.
> 
> I think that this scheme is also sound, but is less precise. E.g., I think it would incorrectly flag the function `ok` in:
> 
> ```
> module M {
>   struct S { f: u64 }
>   spec S { invariant f >= 1 }
> 
>   public fun ok(x: &mut u64, s: &mut S): &mut u64 {
>     s.f += 1;
>     x
>   }
> ```

Correct, however, this is just because the first attempt here is not correct. It does not matter whether the mutated thing is in the return reference. One actually would just need to look if there are any input `&mut`. This would still be not fully precise as it would be now an over-approximation (possibly the function never mutates a parameter), but an analysis complaining anyway is possibly more practical because it is only based on signature and has stable results independent of implementation., In any case, its fine to write up a more precise version, I'm more thinking about what we may actually want to add to the prover.

  
> 
> I don't think this precision is likely to matter much in practice, but I'm somewhat attached to the current approach because we've already formalized it/implemented it/proven it correct + it's also pretty simple.
> 
> > If that is the case, then only data invariants are relevant. Other specs cannot be deferred to the caller (for global invariants, we flag already an error if they are attempting to suspend invariants).
> 
> Yep, I think that makes sense for the current vocabulary of global invariants. To make sure I understand: if we enrich the vocabulary to allow things like `invariant forall s: S: s == S { f: 10 }` (which is currently a dead invariant as we discussed in #9167, but you could imagine supporting it with the natural meaning), or other constructs that quantify over all fields of a given type (e.g., the classic "sum of all `value` fields of `Diem<T>` idea), we would then need to consider global invariants as relevant, right?

This would be not what we understand with global invariants. We should really call this thing "memory invariants" perhaps because its intrinsic part of how we understand them that they only apply to data stored in memory.

What you have in mind is another form of a data invariant, more written like an axiom, but the same would apply regards this discussion as with regular data invariants.

Reply:
In terms of what to actually add to the prover, I was thinking we might need something even more precise than this PR. Specifically, I think we might want to
- track all the access paths `P_leaked` that flow to a mutable return value
- enumerate all the access paths `P_specs` that occur in `struct` specs  in the module and complain if any path in `P_leaked` is a prefix of a path in `P_specs`

This scheme would resolve the false positive reported by this analysis in the `Option` module and a similar one flagged in StarCoin's `Collection2` [code](https://github.com/starcoinorg/starcoin/blob/master/vm/stdlib/modules/Collection2.move#L66).

A more lightweight alternative would be to keep it simple and signature/type-based (as you suggest) and expose a scheme for suppressing spurious warnings like the `Option` one. I think they're infrequent enough that this wouldn't be too annoying.